File: makepro.c
//***************************************************************************
//  MAKEPRO -- Make a Prototype File
//***************************************************************************

/***************************************************************************
    NARS2000 -- An Experimental APL Interpreter
    Copyright (C) 2006-2008 Sudley Place Software

    This program is free software: you can redistribute it and/or modify
    it under the terms of the GNU General Public License as published by
    the Free Software Foundation, either version 3 of the License, or
    (at your option) any later version.

    This program is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU General Public License for more details.

    You should have received a copy of the GNU General Public License
    along with this program.  If not, see <http://www.gnu.org/licenses/>.
***************************************************************************/

#define STRICT
#include <windows.h>
#include <stdio.h>

#include "pcre.h"

#define EQ ==
#define NE !=
typedef char *LPCHAR;
#define Assert(a)     ((a) || (DbgBrk (), DbgBrk (), 0))
void DbgBrk (void);

// Define UBOOL as an unsigned int
typedef unsigned int UBOOL, *LPUBOOL;


//****************************************************************************
//  $IsWhite
//
//  Determine if a char is white space
//****************************************************************************

UBOOL IsWhite
    (char c)

{
    return (c EQ ' ' || c EQ '\t') ? TRUE : FALSE;
} // End IsWhite


//****************************************************************************
//  $SkipWhite
//
//  Skip over leading white space
//****************************************************************************

LPCHAR SkipWhite
    (LPCHAR lp)

{
    // Skip over white space
    while (IsWhite (*lp))
        lp++;

    return lp;
} // End SkipWhite


//***************************************************************************
//  CheckAnachronism
//
//  Check on anachronisms (as the compiler so delicately calls them)
//***************************************************************************

void CheckAnachronism
    (LPCHAR lpHaystack, LPCHAR lpNeedle, LPCHAR lpReplace)

{
    UINT uReplaceLen = lstrlen (lpNeedle);

    while (lpHaystack = strstr (lpHaystack, lpNeedle))
        CopyMemory (lpHaystack, lpReplace, uReplaceLen);
} // End CheckAnachronism


//***************************************************************************
//  DeleteText
//
//  Delete (fill with blanks) given text
//***************************************************************************

void DeleteText
    (LPCHAR lpHaystack, LPCHAR lpNeedle)

{
    UINT uReplaceLen = lstrlen (lpNeedle);

    while (lpHaystack = strstr (lpHaystack, lpNeedle))
        FillMemory (lpHaystack, uReplaceLen, ' ');
} // End DeleteText


//***************************************************************************
//  WinMain
//
//  Start the process
//***************************************************************************

int PASCAL WinMain
    (HINSTANCE   hInstance,
     HINSTANCE   hPrevInstance,     // Zero under Win32
     LPSTR       lpCmdLine,
     int         nCmdShow)

{
    LPCHAR              lpCompiler,
                        lpSwitches,
                        lpOutFile,
                        lpSrcFile,
                        lpTmpFile,
                        lpEndLine;
    SECURITY_ATTRIBUTES secAttr = {sizeof (SECURITY_ATTRIBUTES), NULL, TRUE};
    HANDLE              hOutFile = NULL,
                        hSrcFile = NULL,
                        hTmpFile = NULL,
                        hTmpMap = NULL,
                        hOutMap = NULL,
                        hTmpView = NULL,
                        hOutView = NULL;
    STARTUPINFO         siBlk;
    PROCESS_INFORMATION piBlk;
    DWORD               dwTmpFileSize,
                        dwOutFileSize,
                        dwTempPathLen;
    char                szTempPath[1024];
    UBOOL               bFilesEqual = FALSE;
    FILETIME            CreationTime,
                        LastAccessTime,
                        LastWriteTime;

    // Parse the command line:  {D:\path\to\compiler\cl.exe} {switches} {D:\path\to\sourcefile.ext} {D:\path\to\outputfile.ext}

    lpCompiler = lpCmdLine;

    // Skip over white space
    lpCompiler = SkipWhite (lpCompiler);

    Assert (*lpCompiler EQ '{'); lpCompiler++;
    lpSwitches = strchr (lpCompiler, '}');
    *lpSwitches++ = '\0';

    // Skip over white space
    lpSwitches = SkipWhite (lpSwitches);

    Assert (*lpSwitches EQ '{'); lpSwitches++;
    lpSrcFile = strchr (lpSwitches, '}');
    *lpSrcFile++ = ' ';

    // Skip over white space
    lpSrcFile = SkipWhite (lpSrcFile);

    Assert (*lpSrcFile EQ '{'); *lpSrcFile++ = ' ';
    lpOutFile = strchr (lpSrcFile, '}');
    *lpOutFile++ = '\0';

    // Skip over white space
    lpOutFile = SkipWhite (lpOutFile);

    Assert (*lpOutFile EQ '{'); lpOutFile++;
    lpEndLine = strchr (lpOutFile, '}');
    *lpEndLine++ = '\0';

#if 0
    {
        char szTemp[1024];

        wsprintf (szTemp,
                  "Compiler = <%s>.",
                  lpCompiler);
        MessageBox (NULL, szTemp, "makepro", MB_OK);

        wsprintf (szTemp,
                  "Switches = <%s>.",
                  lpSwitches);
        MessageBox (NULL, szTemp, "makepro", MB_OK);

        wsprintf (szTemp,
                  "SrcFile = <%s>.",
                  lpSrcFile);
        MessageBox (NULL, szTemp, "makepro", MB_OK);

        wsprintf (szTemp,
                  "OutFile = <%s>.",
                  lpOutFile);
        MessageBox (NULL, szTemp, "makepro", MB_OK);
    }
#endif

    // Get a path to temporary files
    dwTempPathLen =
      GetTempPath (sizeof (szTempPath), szTempPath);

    lpTmpFile = &szTempPath[dwTempPathLen + 1];

    // Get a temporary filename
    GetTempFileName (szTempPath, "PRO", 0, lpTmpFile);

    // Create the handle we need for the STARTUPINFO block
    hTmpFile =
      CreateFile (lpTmpFile,                    // Pointer to name of the file
                  GENERIC_READ                  // Access (read-write) mode
                | GENERIC_WRITE,
                  FILE_SHARE_READ
                | FILE_SHARE_WRITE,             // Share mode
                 &secAttr,                      // Pointer to security descriptor
                  CREATE_ALWAYS,                // How to create
                  0                             // File attributes
                | FILE_ATTRIBUTE_TEMPORARY
                | FILE_FLAG_WRITE_THROUGH,
                  NULL);                        // Handle to file with attributes to copy
    if (hTmpFile EQ NULL)
    {
        MessageBox (NULL, "Unable to create hTmpFile", "makepro", MB_OK | MB_ICONEXCLAMATION);
        DbgBrk ();
        GetLastError ();
    } // End IF

    // Set the STARTUPINFO block
    siBlk.cb            = sizeof (STARTUPINFO);
    siBlk.lpReserved    = NULL;
    siBlk.lpDesktop     = NULL;
    siBlk.lpTitle       = "Protoype File Generation";
    siBlk.dwX           = 0;
    siBlk.dwY           = 0;
    siBlk.dwXSize       = 0;
    siBlk.dwYSize       = 0;
    siBlk.dwXCountChars = 0;
    siBlk.dwYCountChars = 0;
    siBlk.dwFlags       = 0
                        | STARTF_USESTDHANDLES
                        | STARTF_USESHOWWINDOW;
    siBlk.wShowWindow   = SW_HIDE;
    siBlk.cbReserved2   = 0;
    siBlk.lpReserved2   = NULL;
    siBlk.hStdInput     = NULL;
    siBlk.hStdOutput    = hTmpFile;
    siBlk.hStdError     = NULL;

    // Create the process to output the prototype
    CreateProcess (lpCompiler,                  // Pointer to name of executable module
                   lpSwitches,                  // Pointer to command line string
                   NULL,                        // Pointer to process security attributes
                   NULL,                        // Pointer to thread security attributes
                   TRUE,                        // Handle inheritance flag
                   0                            // Creationflags
                 | NORMAL_PRIORITY_CLASS,
                   NULL,                        // Pointer to new environment block
                   NULL,                        // Pointer to current directory name
                  &siBlk,                       // Pointer to STARTUPINFO
                  &piBlk);                      // Pointer to PROCESS_INFORMATION
    WaitForSingleObject (piBlk.hProcess, INFINITE);

    // Get the file size (it always fits in a DWORD)
    dwTmpFileSize = GetFileSize (hTmpFile, NULL);
    if (dwTmpFileSize EQ 0)
    {
        hOutFile =
          CreateFile (lpOutFile,                    // Pointer to name of the file
                      GENERIC_READ                  // Access (read-write) mode
                    | GENERIC_WRITE,
                      FILE_SHARE_READ
                    | FILE_SHARE_WRITE,             // Share mode
                     &secAttr,                      // Pointer to security descriptor
                      OPEN_ALWAYS,                  // How to create
                      0                             // File attributes
                    | FILE_ATTRIBUTE_NORMAL
                    | FILE_FLAG_WRITE_THROUGH,
                      NULL);                        // Handle to file with attributes to copy
        if (hOutFile EQ NULL)
        {
            MessageBox (NULL, "Unable to create hOutFile", "makepro", MB_OK | MB_ICONEXCLAMATION);
            DbgBrk ();
            GetLastError ();
        } // End IF

        goto NORMAL_EXIT;
    } // End IF

    // Now launder the output file to remove anomalies of some MS compilers

    // In particular, version 12 of Microsoft Visual Studio mishandles variable arguments
    //   and inserts an extra "..." into the prototype.

    hTmpMap =
      CreateFileMapping (hTmpFile,              // Handle to file to map
                         NULL,                  // Optional security attributes
                         PAGE_READWRITE,        // Protection for mapping object
                         0,                     // High-order 32 bits of object size
                         dwTmpFileSize,         // Low-order 32 bits of object size
                         NULL);                 // Name of file-mapping object
    if (hTmpMap EQ NULL)
    {
        MessageBox (NULL, "Unable to create hTmpMap", "makepro", MB_OK | MB_ICONEXCLAMATION);
        DbgBrk ();
        GetLastError ();
    } // End IF

    hTmpView =
      MapViewOfFile (hTmpMap,                   // File-mapping object to map into address space
                     FILE_MAP_WRITE,            // Access mode
                     0,                         // High-order 32 bits of file offset
                     0,                         // Low-order 32 bits of file offset
                     dwTmpFileSize);            // Number of bytes to map
    if (hTmpView EQ NULL)
    {
        MessageBox (NULL, "Unable to create hTmpView", "makepro", MB_OK | MB_ICONEXCLAMATION);
        DbgBrk ();
        GetLastError ();
    } // End IF

    // Check on unacceptable constructions
    // These constructions, although generated by the compiler as output from /Zg,
    //   are then rejected by it on input.  Go figure.
    CheckAnachronism ((LPCHAR) hTmpView, "... ...",
                                         "...    ");
    CheckAnachronism ((LPCHAR) hTmpView, "__cdecl *__cdecl",
                                         "__cdecl *       ");
    CheckAnachronism ((LPCHAR) hTmpView, "__stdcall *__stdcall",
                                         "__stdcall *         ");
    // Delete some text
    // These statements, although generated by the compiler as output from /Zg,
    //   are then rejected by it on input.  Go figure.
    DeleteText ((LPCHAR) hTmpView, "__forceinline extern void *__cdecl PtrToPtr64(const void *p);");
    DeleteText ((LPCHAR) hTmpView,               "extern void *__cdecl PtrToPtr64(const void *p);");
    DeleteText ((LPCHAR) hTmpView, "__forceinline extern void *__cdecl Ptr64ToPtr(const void *p);");
    DeleteText ((LPCHAR) hTmpView,               "extern void *__cdecl Ptr64ToPtr(const void *p);");
    DeleteText ((LPCHAR) hTmpView, "__forceinline extern void *__cdecl HandleToHandle64(const void *h);");
    DeleteText ((LPCHAR) hTmpView,               "extern void *__cdecl HandleToHandle64(const void *h);");
    DeleteText ((LPCHAR) hTmpView, "__forceinline extern void *__cdecl Handle64ToHandle(const void *h);");
    DeleteText ((LPCHAR) hTmpView,               "extern void *__cdecl Handle64ToHandle(const void *h);");

    // Launder the file to remove extra spaces, CR/LFs, and variable names












    // Now check this file against the previous (if any) output file

    // See if there is a previous copy of the file by attempting to open it
    hOutFile =
      CreateFile (lpOutFile,                    // Pointer to name of the file
                  GENERIC_READ                  // Access (read-write) mode
                | GENERIC_WRITE,
                  FILE_SHARE_READ
                | FILE_SHARE_WRITE,             // Share mode
                 &secAttr,                      // Pointer to security descriptor
                  OPEN_ALWAYS,                  // How to create
                  0                             // File attributes
                | FILE_ATTRIBUTE_NORMAL
                | FILE_FLAG_WRITE_THROUGH,
                  NULL);                        // Handle to file with attributes to copy
    if (hOutFile EQ NULL)
    {
        MessageBox (NULL, "Unable to create hOutFile", "makepro", MB_OK | MB_ICONEXCLAMATION);
        DbgBrk ();
        GetLastError ();
    } // End IF

    // Get the file size (it always fits in a DWORD)
    dwOutFileSize = GetFileSize (hOutFile, NULL);

    // Compare the two file sizes
    if (dwTmpFileSize EQ dwOutFileSize)
    {
        hOutMap =
          CreateFileMapping (hOutFile,              // Handle to file to map
                             NULL,                  // Optional security attributes
                             PAGE_READWRITE,        // Protection for mapping object
                             0,                     // High-order 32 bits of object size
                             dwOutFileSize,         // Low-order 32 bits of object size
                             NULL);                 // Name of file-mapping object
        if (hOutMap EQ NULL)
        {
            MessageBox (NULL, "Unable to create hOutMap #1", "makepro", MB_OK | MB_ICONEXCLAMATION);
            DbgBrk ();
            GetLastError ();
        } // End IF

        hOutView =
          MapViewOfFile (hOutMap,                   // File-mapping object to map into address space
                         FILE_MAP_WRITE,            // Access mode
                         0,                         // High-order 32 bits of file offset
                         0,                         // Low-order 32 bits of file offset
                         dwOutFileSize);            // Number of bytes to map
        if (hOutView EQ NULL)
        {
            MessageBox (NULL, "Unable to create hOutView #1", "makepro", MB_OK | MB_ICONEXCLAMATION);
            DbgBrk ();
            GetLastError ();
        } // End IF

        // Compare the contents
        bFilesEqual =
          memcmp (hTmpView, hOutView, dwOutFileSize) EQ 0;

        // We no longer need this resource
        UnmapViewOfFile (hOutView); hOutView = NULL;

        // We no longer need this resource
        CloseHandle (hOutMap); hOutMap = NULL;
    } // End IF

    // If the files are equal, ...
    if (bFilesEqual)
    {
        // Open the source file (read only)
        hSrcFile =
          CreateFile (lpSrcFile,                    // Pointer to name of the file
                      GENERIC_READ                  // Access (read-write) mode
                    | GENERIC_WRITE,                // Needed for SetFileTime
                      FILE_SHARE_READ
                    | FILE_SHARE_WRITE,             // Share mode
                     &secAttr,                      // Pointer to security descriptor
                      OPEN_EXISTING,                // How to create
                      0                             // File attributes
                    | FILE_ATTRIBUTE_NORMAL
                    | FILE_FLAG_WRITE_THROUGH,
                      NULL);                        // Handle to file with attributes to copy
        if (hSrcFile EQ NULL)
        {
            MessageBox (NULL, "Unable to create hSrcFile", "makepro", MB_OK | MB_ICONEXCLAMATION);
            DbgBrk ();
            GetLastError ();
        } // End IF

        // Copy the file time of the source file (.c) to the output file (.pro)
        //   so we don't run this program again
        GetFileTime (hSrcFile, &CreationTime, &LastAccessTime, &LastWriteTime);
////    SetFileTime (hOutFile, &CreationTime, &LastAccessTime, &LastWriteTime);

        // We no longer need this resource
        CloseHandle (hSrcFile); hSrcFile = NULL;

        // We no longer need this resource
        CloseHandle (hOutFile); hOutFile = NULL;
    } else
    {
        // We no longer need this resource
        CloseHandle (hOutFile); hOutFile = NULL;

        // Create the file as empty
        hOutFile =
          CreateFile (lpOutFile,                    // Pointer to name of the file
                      GENERIC_READ                  // Access (read-write) mode
                    | GENERIC_WRITE,
                      FILE_SHARE_READ
                    | FILE_SHARE_WRITE,             // Share mode
                     &secAttr,                      // Pointer to security descriptor
                      CREATE_ALWAYS,                // How to create
                      0                             // File attributes
                    | FILE_ATTRIBUTE_NORMAL
                    | FILE_FLAG_WRITE_THROUGH,
                      NULL);                        // Handle to file with attributes to copy
        if (hOutFile EQ NULL)
        {
            MessageBox (NULL, "Unable to create hOutFile", "makepro", MB_OK | MB_ICONEXCLAMATION);
            DbgBrk ();
            GetLastError ();
        } // End IF

        // Use the tmp file size
        dwOutFileSize = dwTmpFileSize;

        // Map the file
        hOutMap =
          CreateFileMapping (hOutFile,              // Handle to file to map
                             NULL,                  // Optional security attributes
                             PAGE_READWRITE,        // Protection for mapping object
                             0,                     // High-order 32 bits of object size
                             dwOutFileSize,         // Low-order 32 bits of object size
                             NULL);                 // Name of file-mapping object
        if (hOutMap EQ NULL)
        {
            MessageBox (NULL, "Unable to create hOutMap #2", "makepro", MB_OK | MB_ICONEXCLAMATION);
            DbgBrk ();
            GetLastError ();
        } // End IF

        // View the file
        hOutView =
          MapViewOfFile (hOutMap,                   // File-mapping object to map into address space
                         FILE_MAP_WRITE,            // Access mode
                         0,                         // High-order 32 bits of file offset
                         0,                         // Low-order 32 bits of file offset
                         dwOutFileSize);            // Number of bytes to map
        if (hOutView EQ NULL)
        {
            MessageBox (NULL, "Unable to create hOutView #2", "makepro", MB_OK | MB_ICONEXCLAMATION);
            DbgBrk ();
            GetLastError ();
        } // End IF

        // Copy the tmp file to the out file
        CopyMemory (hOutView, hTmpView, dwTmpFileSize);

        // We no longer need this resource
        UnmapViewOfFile (hOutView); hOutView = NULL;

        // We no longer need this resource
        CloseHandle (hOutMap); hOutMap = NULL;

        // We no longer need this resource
        CloseHandle (hOutFile); hOutFile = NULL;
    } // End IF
NORMAL_EXIT:
    if (hTmpView)
    {
        // We no longer need this resource
        UnmapViewOfFile (hTmpView); hTmpView = NULL;
    } // End IF

    if (hTmpMap)
    {
        // We no longer need this resource
        CloseHandle (hTmpMap); hTmpMap = NULL;
    } // End IF

    if (hOutFile)
    {
        // We no longer need this resource
        CloseHandle (hOutFile); hOutFile = NULL;
    } // End IF

    if (hTmpFile)
    {
        // We no longer need this resource
        CloseHandle (hTmpFile); hTmpFile = NULL;

        // We no longer need this resource
        DeleteFile (lpTmpFile);
    } // End IF

    return 0;
} // End WinMain


//***************************************************************************
//  End of File: makepro.c
//***************************************************************************

File: makepro.php
<?php

/***************************************************************************
    NARS2000 -- An Experimental APL Interpreter
    Copyright (C) 2006-2010 Sudley Place Software

    This program is free software: you can redistribute it and/or modify
    it under the terms of the GNU General Public License as published by
    the Free Software Foundation, either version 3 of the License, or
    (at your option) any later version.

    This program is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU General Public License for more details.

    You should have received a copy of the GNU General Public License
    along with this program.  If not, see <http://www.gnu.org/licenses/>.
***************************************************************************/

// Script to strip out all but prototype text from a .c file
// Date:  15 December 2008
// Author:  Bob Smith


// Parse the argument line
if ($_SERVER['argc'] != 3)
{
    // Display a help message
    echo "Help!  Need two arguments:  typically, foo.c and RET\\ND32\\foo.pro\n";
    exit (1);
} // End IF


// Define <fnmatch> function in case it doesn't exist
if (!function_exists('fnmatch'))
{
    function fnmatch($pattern, $string)
    {
        for ($op = 0, $npattern = '', $n = 0, $l = strlen($pattern); $n < $l; $n++)
        {
            switch ($c = $pattern[$n])
            {
                case '\\':
                    $npattern .= '\\' . @$pattern[++$n];

                    break;

                case '.':
                case '+':
                case '^':
                case '$':
                case '(':
                case ')':
                case '{':
                case '}':
                case '=':
                case '!':
                case '<':
                case '>':
                case '|':
                    $npattern .= '\\' . $c;

                    break;

                case '?':
                case '*':
                    $npattern .= '.' . $c;

                    break;

                case '[':
                case ']':
                default:
                    $npattern .= $c;
                    if ($c == '[')
                    {
                        $op++;
                    } else
                    if ($c == ']')
                    {
                        if ($op == 0)
                            return false;
                        $op--;
                    } // End IF/ELSE

                    break;
            } // End SWITCH
        } // End FOR

        if ($op != 0) return false;
            return preg_match('/^' . $npattern . '/i', $string);
    } // End fnmatch
} // End IF



//****************************************************//
// Get the command line arguments

$FileWild = $_SERVER['argv'][1];    // Get the input file (possibly with wildcards)
$OutFile  = $_SERVER['argv'][2];    // Get the output file

// Split apart the file directory name and base
$FileDir  = dirname  ($FileWild);
$FileBase = basename ($FileWild);

$hDir = opendir ($FileDir);

// If the file dir doesn't end with a backslash, do so now
if (substr ($FileDir, -1, 1) != "\\")
    $FileDir .= "\\";

// Loop through the files in the dir
while (($FileName = readdir ($hDir)) !== false)
{
    if ($FileName != '.'
     && $FileName != '..'
     && $FileName != 'makever.c')
    {
        if (fnmatch ($FileBase, $FileName)
         && ($FileName == 'bjhash.c'
          || $FileName == 'dtoa.c'
          || $FileName == 'cs_parse.c'
          || $FileName == 'fh_parse.c'
          || $FileName == 'pl_parse.c'
          || $FileName == 'pn_parse.c'))
        {
            // Copy the .proto file from the source dir to the target dir
            $InpFile = basename ($FileName, ".c") . ".proto";

            // Copy the file
            copy ($InpFile, $OutFile);
        } else
        if (fnmatch ($FileBase, $FileName))
        {
            $path_parts = pathinfo ($FileName);
            if (strtolower ($path_parts['extension']) == 'c')
                ProcessFile ($FileDir . $FileName, $OutFile);
        } // End IF/ELSE/...
    } // End IF
} // End WHILE

closedir ($hDir);

exit (0);

// Function to process a single file
function ProcessFile ($FileName, $OutFile)
{
    // Read in the file contents
    $FileNew = file_get_contents ($FileName);
    if ($FileNew == false)
    {
        echo "Unable to read the filename <$FileName>\n";
        exit (2);
    } // End IF

    // In case this is a .c file from a .y file, replace single \n with \r\n
    $count = 1;
    while ($count > 0)
        $FileNew = preg_replace ('/([^\r])\n/', "$1\r\n", $FileNew, -1, $count);

    // Strip out single line comments
    $FileNew = preg_replace ('/\/\/.*?\r\n/s', "\r\n", $FileNew);

    // Strip out multiple line comments
    $FileNew = preg_replace ('/\/\*.*?\*\//s', '', $FileNew);

    // Strip out character strings
    $FileNew = preg_replace ('/".*?"/', '', $FileNew);
    $FileNew = preg_replace ('/\'.*?\'/', '', $FileNew);

    // Strip out txt between braces
    $count = 1;
    while ($count > 0)
        $FileNew = preg_replace ('/{[^{]*?}/s', '', $FileNew, -1, $count);

    // Merge continuation lines
    $count = 1;
    while ($count > 0)
        $FileNew = preg_replace ('/\x5c\r\n/m', '', $FileNew, -1, $count);

    // Strip out #include lines
    $FileNew = preg_replace ('/#\s*include.*?\r\n/', '', $FileNew);

    // Strip out #pragma lines
    $FileNew = preg_replace ('/#\s*pragma.*?\r\n/', '', $FileNew);

    // Strip out #define .* lines
    $FileNew = preg_replace ('/#\s*define .*?\r\n/', '', $FileNew);

    // Strip out typedef {.*}; lines
    $FileNew = preg_replace ('/typedef .*?{.*?};\r\n/s', '', $FileNew);

    // Strip out typedef .*; lines
    $FileNew = preg_replace ('/typedef .*?;\r\n/s', '', $FileNew);

    // Strip out #undef .* lines
    $FileNew = preg_replace ('/#\s*undef .*?\r\n/', '', $FileNew);

    // Strip out static .*?; lines
    $FileNew = preg_replace ('/static.*?;\r\n/', '', $FileNew);

    // Strip out static (.*) lines
    $FileNew = preg_replace ('/static.*?\(.*?\)\r\n/ms', '', $FileNew);

    // Strip out blank lines
    $FileNew = StripBlankLines ($FileNew);

    // Strip out empty #if... #endif statements
    $FileNew = StripEmptyDecl ($FileNew);

    // Strip out trailing blanks on a line
    $FileNew = preg_replace ('/\s+\r\n/', "\r\n", $FileNew);

    // Merge lines that do not begin with a '#' and end with a comma or equal sign
    //   with the next line if it does not begin with a '#'
    $count = 1;
    while ($count > 0)
        $FileNew = preg_replace ('/^([^#].*?)([,=])\r\n([^#])/m', '$1$2$3', $FileNew, -1, $count);

    // Replace ')\r\n' with ');\r\n'
    $FileNew = preg_replace ('/\)\r\n/', ");\r\n", $FileNew);

    // Replace '\r\n\s+\(' with "\r\n("
    $FileNew = preg_replace ('/\r\n\s+\(/', "\r\n(", $FileNew);

    // Strip out spurious blanks
    $count = 1;
    while ($count > 0)
        $FileNew = preg_replace ('/  /', ' ', $FileNew, -1, $count);

    // Merge lines that do not begin with '#' or '(' with a next line
    //   that does begin with '('
    $FileNew = preg_replace ('/^([^#(].*?)\r\n\(/m', '$1 (', $FileNew);

    // Strip out lines that do not begin with '#' and have no parens
    $FileNew = preg_replace ('/^[^#()]+;/m', '', $FileNew);

    // Strip out empty #if... #endif statements
    $FileNew = StripEmptyDecl ($FileNew);

    // Strip out blank lines
    $FileNew = StripBlankLines ($FileNew);

    // Strip out '#ifdef .*?\r\n#endif\r\n'
    $FileNew = preg_replace ('/^#ifdef .*?\r\n#endif\r\n/m', '', $FileNew);

    // Strip out '#ifndef PROTO.*?#endif\r\n'
    $FileNew = preg_replace ('/^#ifndef PROTO.*?^#endif\r\n/ms', '', $FileNew);

    // Strip out '#ifdef __cplusplus.*?#endif\r\n'
    $FileNew = preg_replace ('/^#ifdef __cplusplus.*?#endif\r\n/ms', '', $FileNew);

    // Strip out blank lines
    $FileNew = StripBlankLines ($FileNew);

    // Replace "\r\n;" with ";"
    $FileNew = preg_replace ('/\r\n;/', ';', $FileNew);

    // In lines that begin with a '#' and end with ");", replace the ");" with ")"
    $FileNew = preg_replace ('/(#.*?\));\r\n/', "$1\r\n", $FileNew);

    // Strip out lines that do not begin with '#' and have no parens
    $FileNew = preg_replace ('/^[^#()]+;/m', '', $FileNew);

    // Strip out leading \r\n
    while (substr ($FileNew, 0, 2) == "\r\n")
        $FileNew = substr ($FileNew, 2);

    // Get the resulting length
    $FileLen = strlen ($FileNew);

    if (file_exists ($OutFile))
        // Read in the existing file contents
        $FileOld = file_get_contents ($OutFile);
    else
        $FileOld = false;

    // Compare against the existing file:
    //   if the same, don't write out (and don't change the timestamp)
    if ($FileOld != false
    && $FileLen == strlen ($FileOld)
    && strncmp ($FileNew, $FileOld, $FileLen) == 0)
        return;

    // Write out the file contents
    file_put_contents ($OutFile, $FileNew);

////echo "Output file written to <$OutFile>.\n";
} // End ProcessFile


// Strip out empty declarations
function StripEmptyDecl ($File)
{
    $count1 = $count2 = 1;
    while ($count1 > 0 || $count2 > 0)
    {
        // Strip out '#if.*?(\r\n)+#endif\r\n'
        $File = preg_replace ('/^#\s*if.*?(\r\n\s*)+#\s*endif(\r\n\s*)+/m', '', $File, -1, $count1);

        // Strip out #if...(#elif...)#else...#endif
        $File = preg_replace ('/^#\s*if.*?(\r\n\s*)+(#\s*elif.*?(\r\n\s*)+)*#\s*else(\r\n\s*)+#\s*endif(\r\n\s*)+/m', '', $File, -1, $count2);
    } // End WHILE

    return $File;
} // End StripEmptyDecl


// Strip out blank lines
function StripBlankLines ($File)
{
    $count = 1;
    while ($count > 0)
        $File = preg_replace ('/\r\n\r\n/', "\r\n", $File, -1, $count);

    return $File;
} // End StripBlankLines

?>
